src/kaleidoscope/device/dygma/Raise.cpp

```diff
@@ -0,0 +1,599 @@
+uint16_t RaiseHands::keyscan_interval_ = 50;
+
+void RaiseHands::setSidePower(bool power) {
+  digitalWrite(SIDE_POWER, power ? HIGH : LOW);
+  side_power_ = power;
+}
+
+void RaiseHands::setup() {
+  settings_base_ = ::EEPROMSettings.requestSlice(sizeof(keyscan_interval_));
+
+  // If keyscan is max, assume that EEPROM is uninitialized, and store the
+  // defaults.
+  uint16_t interval;
+  Kaleidoscope.storage().get(settings_base_, interval);
+  if (interval == 0xffff) {
+    Kaleidoscope.storage().put(settings_base_, keyscan_interval_);
+    Kaleidoscope.storage().commit();
+  }
+  Kaleidoscope.storage().get(settings_base_, keyscan_interval_);
+}
+
+void RaiseHands::keyscanInterval(uint16_t interval) {
+  leftHand.setKeyscanInterval(interval);
+  rightHand.setKeyscanInterval(interval);
+  keyscan_interval_ = interval;
+  Kaleidoscope.storage().put(settings_base_, keyscan_interval_);
+  Kaleidoscope.storage().commit();
+}
+
+void RaiseHands::initializeSides() {
+  // key scan interval from eeprom
+  leftHand.setKeyscanInterval(keyscan_interval_);
+  rightHand.setKeyscanInterval(keyscan_interval_);
+
+  // get ANSI/ISO at every side replug
+  uint8_t l_ansi_iso = leftHand.readANSI_ISO();
+  uint8_t r_ansi_iso = rightHand.readANSI_ISO();
+
+  // setup ansi_iso variable, this will affect led mapping - defaults to ISO if nothing reported
+  if (l_ansi_iso == ANSI || r_ansi_iso == ANSI)
+    ansi_iso = ANSI;
+  else
+    ansi_iso = ISO;
+
+  /*
+   * if the neuron starts up with no sides connected, it will assume ISO. This
+   * turns on an extra LED (hardware LED 19 on left side). If an ANSI left is
+   * then plugged in, the keyboard will switch to ANSI, but LED 19 can't get
+   * wiped because the ANSI LED map doesn't include this LED. It will be driven
+   * from the SLED1735's memory with the same colour as before, which causes
+   * weird looking colours to come on on other seemingly unrelated keys. So: on
+   * a replug, set LED 19 to off to be safe.
+   */
+  leftHand.led_data.leds[19] = {0, 0, 0};
+
+  // get activated LED plugin to refresh
+  ::LEDControl.refreshAll();
+}
+
+/********* LED Driver *********/
+
+bool RaiseLEDDriver::isLEDChangedNeuron;
+uint8_t RaiseLEDDriver::isLEDChangedLeft[LED_BANKS];
+uint8_t RaiseLEDDriver::isLEDChangedRight[LED_BANKS];
+cRGB RaiseLEDDriver::neuronLED;
+
+void RaiseLEDDriver::syncLeds() {
+  // left and right sides
+  for (uint8_t i = 0; i < LED_BANKS; i ++) {
+    // only send the banks that have changed - try to improve jitter performance
+    if (isLEDChangedLeft[i]) {
+      RaiseHands::leftHand.sendLEDBank(i);
+      isLEDChangedLeft[i] = false;
+    }
+    if (isLEDChangedRight[i]) {
+      RaiseHands::rightHand.sendLEDBank(i);
+      isLEDChangedRight[i] = false;
+    }
+  }
+
+  if (isLEDChangedNeuron) {
+    updateNeuronLED();
+    isLEDChangedNeuron = false;
+  }
+}
+
+void RaiseLEDDriver::updateNeuronLED() {
+  static constexpr struct {
+    uint8_t r, g, b;
+  } pwm = { 3, 5, 4 };
```

Comment:
What are the 3,4,5?

Reply:
Those are the Neuron's LED pins, I think. The Neuron is the brain of the Raise, and it has a single LED on it, which this function updates to the desired color. This is something that is - by design - not under LEDControl's command, hence the separate function.